src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs

```diff
@@ -204,33 +202,57 @@ private SingleTypeDeclaration FirstDeclarationWithExplicitBases()
             return null;
         }
 
-        internal Tuple<NamedTypeSymbol, ImmutableArray<NamedTypeSymbol>> GetDeclaredBases(ConsList<Symbol> basesBeingResolved)
+        /// <summary>
+        /// ignoreNonNullTypesAttribute is used when we need to break cycles
+        /// </summary>
+        internal Tuple<NamedTypeSymbol, ImmutableArray<NamedTypeSymbol>> GetDeclaredBases(ConsList<Symbol> basesBeingResolved, bool ignoreNonNullTypesAttribute)
         {
             if (ReferenceEquals(_lazyDeclaredBases, null))
             {
-                DiagnosticBag diagnostics = DiagnosticBag.GetInstance();
-                if (Interlocked.CompareExchange(ref _lazyDeclaredBases, MakeDeclaredBases(basesBeingResolved, diagnostics), null) == null)
+                initDeclaredBasesIgnoringNonNullTypesAttributeIfNeeded();
+                if (ignoreNonNullTypesAttribute)
                 {
-                    AddDeclarationDiagnostics(diagnostics);
+                    return _lazyDeclaredBasesIgnoringNonNullTypesAttribute;
                 }
 
-                diagnostics.Free();
+                // We already computed bases ignoring NonNullTypes (ie. assuming NonNullTypes is true),
+                // so diagnostics were already reported and we can just fix the base and interfaces up with NonNullTypes information
+                var lazyDeclaredBases = NonNullTypes ?
+                    _lazyDeclaredBasesIgnoringNonNullTypesAttribute :
+                    new Tuple<NamedTypeSymbol, ImmutableArray<NamedTypeSymbol>>(
+                        (NamedTypeSymbol)_lazyDeclaredBasesIgnoringNonNullTypesAttribute.Item1?.SetUnknownNullabilityForReferenceTypes(),
+                        _lazyDeclaredBasesIgnoringNonNullTypesAttribute.Item2.SelectAsArray(i => (NamedTypeSymbol)i.SetUnknownNullabilityForReferenceTypes()));
+
+                Interlocked.CompareExchange(ref _lazyDeclaredBases, lazyDeclaredBases, null);
             }
 
             return _lazyDeclaredBases;
+
+            void initDeclaredBasesIgnoringNonNullTypesAttributeIfNeeded()
```

Comment:
>initDeclaredBasesIgnoringNonNullTypesAttributeIfNeeded [](start = 17, length = 54)

Consider inlining this local function. There's only one caller and, as a reader, I'd expect to read this code before reading the caller above.